File: solns/scala/28.scala
def turn(heading: Symbol) = heading match {
  case 'right => 'down
  case 'down  => 'left
  case 'left  => 'up
  case 'up    => 'right
}

def next(x: Int, y: Int, heading: Symbol) = heading match {
  case 'right => (x + 1, y)
  case 'down  => (x, y - 1)
  case 'left  => (x - 1, y)
  case 'up    => (x, y + 1)
}

def populate(bound: Int) = {
  val m = Array.ofDim[Int](bound, bound)
  def bounded(p: Tuple2[Int, Int]) = {
    val (x, y) = p
    x >= 0 && x < m.length && y >= 0 && y < m.length
  }
  def recur(p: Tuple2[Int, Int], heading: Symbol, n: Int): Unit = {
    if (bounded(p)) {
      val (x, y) = p
      val (u, v) = next(x, y, turn(heading))
      m(x)(y) = n
      m(u)(v) match {
        case 0 => recur((u, v), turn(heading), n + 1)
        case _ => recur(next(x, y, heading), heading, n + 1)
      }
    }
  }
  recur((m.length / 2, m.length / 2) /* p */, 'up /* heading */, 1 /* n */)
  m
}

val b = 1001
val m = populate(b)
// Sum the diagonals but subtract 1 because it's counted twice
val result = (0 until b)
    .map(k => m(k)(k) + m(k)(m.length - k - 1))
    .reduceLeft(_ + _) - 1

println(result)
assert(result == 669171001)

File: project_euler/28.py
# Problem 28: Number Spiral Diagonals
#
# What is the sum of the numbers on the diagonals in a 1001x1001 spiral formed in the same way?
#
# See https://projecteuler.net/problem=28
import numpy as np


def turn(direction: str) -> str:
    return {"R": "D", "D": "L", "L": "U", "U": "R"}[direction]


def next(position: tuple[int, int], direction: str) -> tuple[int, int]:
    i, j = position

    return {"R": (i + 1, j), "D": (i, j - 1), "L": (i - 1, j), "U": (i, j + 1)}[direction]


def create_spiral(n: int) -> np.ndarray:
    spiral = np.zeros((n, n), dtype=int)

    mid = n // 2
    heading = "U"
    (i, j) = (mid, mid)

    for x in range(1, n * n + 1):
        # NumPy is row-major order; that is spiral[i][j] == spiral[j, i].
        spiral[j, i] = x
        next_heading = turn(heading)
        (ni, nj) = next((i, j), next_heading)

        if 0 <= ni < n and 0 <= nj < n and spiral[nj, ni] == 0:
            heading = next_heading
            (i, j) = (ni, nj)
        else:
            (i, j) = next((i, j), heading)

    return spiral


def run(limit: int) -> int:
    spiral = create_spiral(limit)
    # Compute sum from top left to bottom right.
    main_diag = np.trace(spiral)
    # Compute sum from bop right to bottom left but horizontally flipping the spiral.
    anti_diag = np.trace(np.fliplr(spiral))
    # Center is double counted, so subtract 1.
    return int(main_diag + anti_diag - 1)


def test_run():
    assert run(1001) == 669171001
